airbyte-integrations/bases/standard-test/standard_test/standard_test.py

```diff
@@ -0,0 +1,84 @@
+"""
+MIT License
+
+Copyright (c) 2020 Airbyte
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from collections import Counter
+
+import inflection as inflection
+import pytest
+from airbyte_protocol import Type, Status, ConnectorSpecification
+from .utils import full_refresh_only_catalog, ConnectorRunner
+
+
+@pytest.mark.usefixtures("inputs")
+class BaseTest:
+
+    @classmethod
+    def config_key(cls):
+        class_name = cls.__name__
+        if class_name.startswith("Test"):
+            class_name = class_name[len("Test"):]
+        return inflection.underscore(class_name)
+
+
+class TestCore(BaseTest):
+    def test_check(self, connector_config, docker_runner: ConnectorRunner):
+        output = docker_runner.call_check(config=connector_config)
+        con_messages = [message for message in output if message.type == Type.CONNECTION_STATUS]
+
+        assert len(con_messages) == 1, "Connection status message should be emitted exactly once"
+        assert con_messages[0].connectionStatus.status == Status.SUCCEEDED
+
+    def test_discover(self, connector_config, catalog, docker_runner: ConnectorRunner):
+        output = docker_runner.call_discover(config=connector_config)
+        catalog_messages = [message for message in output if message.type == Type.CATALOG]
+
+        assert len(catalog_messages) == 1, "Catalog message should be emitted exactly once"
+        # assert catalog_messages[0].catalog == catalog, "Catalog should match the one that was provided"
+
+    def test_spec(self, connector_spec, docker_runner: ConnectorRunner):
+        output = docker_runner.call_spec()
+        spec_messages = [message for message in output if message.type == Type.SPEC]
+
+        assert len(spec_messages) == 1, "Spec message should be emitted exactly once"
+        # assert spec_messages[0].spec == spec, "Spec should be equal to the one in spec.json file"
+
+    def test_read(self, connector_config, configured_catalog, docker_runner: ConnectorRunner):
```

Comment:
not required for this PR (we should create an issue) but how can we dynamically export the test cases into a readable format like [this](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/standard-source-tests) (or even more readable 😉 ) . 

The current logic in the java test suite runs [as part of CI](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/standard-source-test/build.gradle#L36) and reads the test source file, parses the java docs for it, and dynamically generates the markdown. This way, there is a very tight coupling between the docs produced for the user and the living, breathing test case code. I admit this is not a super "elegant" way of doing it but it does achieve the desired outcome. WDYT?

Reply:
Well, the right way to do is with sphinx plugin as described [here](https://stackoverflow.com/a/58929446/656671).
But it is really easy to just parse docstrings with python and generate this kind of page.
The thing is that the output from the test is quite detailed, and can be customized with plugins like [this one](https://pypi.org/project/pytest-spec/) and [this one](https://pypi.org/project/pytest-testdox/)
so when test fails we have this description of its logic in the output.
Another place to put it is config schema (pydantic model) so when we generate schema from it we can describe each test.

Reply:
btw, I think https://github.com/airbytehq/airbyte/issues/2182 covers this

Reply:
yup, agreed. We can punt this to #2182